packages/go_router/lib/src/configuration.dart

```diff
@@ -111,6 +124,90 @@ class RouteConfiguration {
     return true;
   }
 
+  // Check to see that the configured initialLocation of StatefulShellBranches
+  // points to a descendant route of the route branch.
+  bool _debugCheckStatefulShellBranchDefaultLocations(
+      List<RouteBase> routes, RouteMatcher matcher) {
+    try {
+      for (final RouteBase route in routes) {
+        if (route is StatefulShellRoute) {
+          for (final StatefulShellBranch branch in route.branches) {
+            if (branch.initialLocation == null) {
+              // Recursively search for the first GoRoute descendant. Will
+              // throw assertion error if not found.
+              findStatefulShellBranchDefaultLocation(branch);
+            } else {
+              final RouteBase initialLocationRoute =
+                  matcher.findMatch(branch.initialLocation!).last.route;
+              final RouteBase? match = branch.routes.firstWhereOrNull(
+                  (RouteBase e) => _debugIsDescendantOrSame(
+                      ancestor: e, route: initialLocationRoute));
+              assert(
+                  match != null,
+                  'The initialLocation (${branch.initialLocation}) of '
+                  'StatefulShellBranch must match a descendant route of the '
+                  'branch');
+            }
+          }
+        }
+        _debugCheckStatefulShellBranchDefaultLocations(route.routes, matcher);
+      }
+    } on MatcherError catch (e) {
+      assert(
+          false,
+          'initialLocation (${e.location}) of StatefulShellBranch must '
+          'be a valid location');
+    }
+    return true;
+  }
+
+  static Iterable<RouteBase> _subRoutesRecursively(List<RouteBase> routes) =>
+      routes.expand(
+          (RouteBase e) => <RouteBase>[e, ..._subRoutesRecursively(e.routes)]);
+
+  static GoRoute? _findFirstGoRoute(List<RouteBase> routes) =>
+      _subRoutesRecursively(routes)
+          .firstWhereOrNull((RouteBase e) => e is GoRoute) as GoRoute?;
```

Comment:
`whereType<GoRoute>().firstOrNull()`

Reply:
Although this way of writing is more readable, but it needs to traverse all the data.
Possibly worse performance instead?

Reply:
@tatsuyuki25 the code I wrote does exactly the same amount of operations (if not less) than the initial code

Reply:
I read the code of the collection package, it is indeed the same amount of operations.

Reply:
Great readability refinement , @ahmednfwela, thanks. I think I’ll add this, even though (to nitpick in the extreme) it has ever so slightly worse performance, since it involves creating a wrapping/nested `Iterable`, whereas `firstWhereOrNull` only uses a simple for-in loop. But the real performance "villan" here is of course _subRoutesRecursively, will adress that too.